src/python/content.cpp

```diff
@@ -1676,6 +1734,52 @@ make_ListOffsetArrayOf(const py::handle& m, const std::string& name);
 
 ////////// NumpyArray
 
+const ak::NumpyArray
+NumpyArray_from_cupy(const std::string& name,
+                     const py::object& array,
+                     const py::object& identities,
+                     const py::object& parameters) {
```

Comment:
This was split out so that it can be used in the NumpyArray constructor as well, and so a lot of places in the Python code don't have to be changed to "if CuPy, call `NumpyArray.from_cupy`, else call the constructor." This is a case of hiding complexity so that more code can be naive.